### PR TITLE
Add functionality to persist descriptions to nested columns in bigquery

### LINKED ISSUE #2549: persist_docs functionality doesn't persist descriptions on nested bigquery columns
### Describe the bug
With the release of dbt 0.17.0, the persist_docs functionality was enhanced to also apply column descriptions to bigquery when they are supplied in a schema.yml file. However, it seems that it only takes into account top-level columns and there is no way to persist a description to a nested column (ie a field that is part of a RECORD type bigquery column).

### Steps To Reproduce
Suppose you create a simple test model using dbt in bigquery:
``` SQL
SELECT
    STRUCT(
        STRUCT(
            1 AS level_3_a,
            2 AS level_3_b
        ) AS level_2
    ) AS level_1
```

This will create a view/table in bigquery with the following schema:
image

We can describe this table in a schema.yml file and add descriptions as follows:
```yaml
version: 2

models:
  - name: test_nested
  
    columns:
      - name: level_1
        description: level_1 description
      - name: level_1.level_2.level_3_a
        description: level_3_a description
```

When we now perform a `dbt run` on this model, only the description on the top level field is added and no warnings or errors are generated by dbt:
image

### Expected behavior
When providing descriptions for nested fields (preferably using the dotted syntax as shown above), the descriptions should be correctly persisted on nested fields.

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [ ] redshift
- [x] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
installed version: 0.17.0
   latest version: 0.17.0

Up to date!

Plugins:
  - snowflake: 0.17.0
  - postgres: 0.17.0
  - bigquery: 0.17.0
  - redshift: 0.17.0
```

**The operating system you're using:**
Mac OS 10.15.3

**The output of `python --version`:**
Python 3.7.5

### PR DESCRIPTION
**Pull request by bodschut**
resolves #2549 

### Description

Changed the existing function in the bigquery adapter implementation to recursively traverse nested columns in bigquery and add descriptions if they are provided in a schema.yml file. Requires that nested fields are addressed with dotted syntax in the schema file.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by drewbanin** [comment]
thanks for opening this PR @bodschut! This is a really good and positive change :)

Are you able to sign the [CLA form](https://docs.getdbt.com/docs/contributing/contributor-license-agreements/)? Once that's done, we'd be happy to review this PR!

**Comment by bodschut** [comment]
Ok, I signed it :-)

**Comment by drewbanin** [comment]
@cla-bot check

**Comment by drewbanin** [comment]
ok - thanks for signing it! I do see the signature, but the CLA bot is upset that your existing commit is not associated with your GitHub user :). This is the mechanism used to determine that commits to this repo are contributed by CLA signatories, so if you could try running
```
git config --global user.email email@example.com
```

with your github email address, i think we should be in good shape.

We'll take a look at this PR and add some comments inline! Thanks again for opening it :)

**Comment by bodschut** [comment]
Yeah, I signed the CLA with my personal email address and the commit was probably done with my professional address... I changed it locally so I think I should do a dummy commit to get it in?

**Comment by beckjake** [comment]
I think you'll want to `git commit --amend --reset-author --no-edit` and then force-push once you fix the `git config` settings.

**Comment by beckjake** [comment]
You could also add your professional address to your github, if that works for you. The cla-bot actually goes by username, and the problem is that github doesn't know that your commit is from `@bodschut` because it just compares emails (I think!).

**Comment by bodschut** [comment]
seems to be resolved now, thanks for the help :-)

File: plugins/bigquery/dbt/adapters/bigquery/impl.py [hunk 1]
```diff
@@ -595,6 +595,19 @@ def get_table_ref_from_relation(self, conn, relation):
                                           relation.identifier,
                                           conn)
 
+    def _update_column_dict(self, bq_column_dict, dbt_columns, parent=''):
+        # bq_colunmn_dict generated by SchemaField.to_api_repr()
+        dotted_column_name = '{}.{}'.format(parent, bq_column_dict['name']) if parent else bq_column_dict['name']
```
Comment (beckjake) [inline plugins/bigquery/dbt/adapters/bigquery/impl.py:600]: Please split this up into a normal multiline if-else statement, this is too complicated for a one-liner.

File: plugins/bigquery/dbt/include/bigquery/macros/catalog.sql [hunk 2]
```diff
@@ -99,10 +99,11 @@
             concat(table_catalog, '.', table_schema, '.', table_name) as relation_id,
             field_path as column_name,
             data_type as column_type,
-            column_name as base_column_name
+            column_name as base_column_name,
+            description as column_comment
 
         from {{ information_schema.replace(information_schema_view='COLUMN_FIELD_PATHS') }}
-        where data_type not like 'STRUCT%'
+        -- where data_type not like 'STRUCT%'
```
Comment (beckjake) [inline plugins/bigquery/dbt/include/bigquery/macros/catalog.sql:106]: Please just remove this line - no need for commenting out code!
```suggestion
```

File: plugins/bigquery/dbt/include/bigquery/macros/catalog.sql [hunk 3]
```diff
@@ -83,7 +83,7 @@
             -- use the "real" column name from the paths query below
             column_name as base_column_name,
             ordinal_position as column_index,
-            cast(null as string) as column_comment,
+            -- cast(null as string) as column_comment,
```
Comment (beckjake) [inline plugins/bigquery/dbt/include/bigquery/macros/catalog.sql:86]: Please just remove this line - no need for commenting out code!
```suggestion
```

Comment (bodschut) [reply plugins/bigquery/dbt/include/bigquery/macros/catalog.sql:86]: yep sorry, forgot to remove it :-)

**Comment by beckjake** [comment]
This is a weird result! The indices for the nested structs seem to have changed as a result of this PR - this is the part of the diff that actually reflects a failure:
```
E   -                                         'nested_field': {'comment': None,
 
E   -                                                          'index': 4,
 
E   -                                                          'name': 'nested_field',
 
E   -                                                          'type': 'STRUCT<field_4 '
 
E   -                                                                  'INT64, '
 
E   -                                                                  'field_5 '
 
E   -                                                                  'INT64>'},
 
E                                             'nested_field.field_4': {'comment': None,
 
E   -                                                                  'index': 5,
 
E   ?                                                                           ^
 
E   
 
E   +                                                                  'index': 4,
 
E   ?                                                                           ^
 
E   
 
E                                                                      'name': 'nested_field.field_4',
 
E                                                                      'type': 'INT64'},
 
E                                             'nested_field.field_5': {'comment': None,
 
E   -                                                                  'index': 6,
 
E   ?                                                                           ^
 
E   
 
E   +                                                                  'index': 5,
 
E   ?                                                                           ^
 
E   
 
E                                                                      'name': 'nested_field.field_5',
 
E                                                                      'type': 'INT64'}},
 
E                                 'metadata': {'comment': None,
 ```

**Comment by bodschut** [comment]
Bigquery allows you to put a description on the parent field that has type STRUCT, that's why I removed the where clause that excluded these. Maybe we should take that into account in the test?

**Comment by drewbanin** [comment]
This is awesome!! Longer term, I think I'd like the docs site to render these fields more like the BQ UI:
Screen Shot 2020-06-19 at 3 52 04 PM

That said, I think this PR is great, and we'll have all the info we need in the docs site to implement those changes when we choose to. Approved on my end :)

**Pull request merged**